pyuvdata/utils.py

```diff
@@ -492,6 +492,91 @@ def uvcalibrate(uvdata, uvcal, inplace=True, prop_flags=True, flag_missing=True,
         return uvdata
 
 
+def apply_uvflag(uvd, uvf, inplace=True, force_pol=True, unflag_first=False,
+                 flag_missing=True):
+    """
+    Apply flags from a UVFlag to a UVData instantiation. This edits inplace by default.
+
+    Parameters
+    ----------
+    uvd : UVData object
+        UVData object to add flags to.
+    uvf : UVFlag object
+        A UVFlag object in flag mode.
+    inplace : bool
+        If True overwrite flags in uvd, otherwise return new object
+    force_pol : bool
+        If True, broadcast flags to all polarizations if they do not match.
+        Only works if uvf.Npols == 1.
+    unflag_first : bool
+        If True, completely unflag the UVData before applying flags.
+        Else, OR the inheret uvd flags with uvf flags.
+    flag_missing : bool
+        If input uvf is a baseline type and antpairs in uvd do not exist in uvf,
+        flag them in uvd. Otherwise leave them untouched.
+
+    Returns
+    -------
+    UVData
+        If not inplace, returns new UVData object with flags applied
+    """
+    # assertions
+    if uvf.mode != 'flag':
+        raise ValueError("UVFlag must be flag mode")
+
+    if not inplace:
+        uvd = copy.deepcopy(uvd)
+
+    # make a deepcopy by default b/c it is generally edited inplace downstream
+    uvf = copy.deepcopy(uvf)
+
+    # convert to baseline type
+    if uvf.type != 'baseline':
+        # edits inplace
+        uvf.to_baseline(uvd, force_pol=force_pol)
+
+    else:
+        # make sure polarizations match or force_pol
+        uvd_pols, uvf_pols = uvd.polarization_array.tolist(), uvf.polarization_array.tolist()
+        if set(uvd_pols) != set(uvf_pols):
+            if uvf.Npols == 1 and force_pol:
+                # if uvf is 1pol we can make them match: also edits inplace
+                uvf.polarization_array = np.repeat(uvf.polarization_array, uvd.Npols, axis=-1)
+                uvf.Npols = len(uvf.polarization_array)
+                uvf_pols = uvf.polarization_array.tolist()
+
+            else:
+                raise ValueError("Input uvf and uvd polarizations do not match")
+
+        # make sure polarization ordering is correct: also edits inplace
+        uvf.polarization_array = uvf.polarization_array[[uvd_pols.index(pol) for pol in uvf_pols]]
+
+        # check time and freq shapes match: if Ntimes or Nfreqs is 1, allow implicit broadcasting
+        if uvf.Ntimes != uvd.Ntimes and uvf.Ntimes > 1:
+            raise ValueError("UVFlag and UVData have mismatched Ntimes")
+        if uvf.Nfreqs != uvd.Nfreqs and uvf.Nfreqs > 1:
+            raise ValueError("UVFlag and UVData have mismatched Nfreqs")
```

Comment:
I think we should be checking for overlapping times and frequencies, not that they are the same size. With this check we could have the same length of observations with a time translation and the flags would still be applied.

Reply:
I don't actually think that's necessary. I do agree that it is true one can do that, but I don't think we should be in a position to tell the user that its the wrong thing to do: I can think of some reasons why one might want to knowingly do something like that. The check we do have (above) is more of a practical one: when Ntimes and Nfreqs don't match, you simply cannot add two arrays together, so its more of an elegant failure, rather than telling the user they can't do something.

Reply:
I can see the advantage to that, maybe there can be a flag to force time/frequency mismatches? It feels like the base use for this kind of function would be: turn metrics from a observation into flags, apply those flags to the same data the metrics/etc came from. From that perspective it feels like a check should be made that they match.  I think this would more elegantly do what @adampbeardsley suggested. Check for overlap, if no overlap and flag_anyway and size are same do it.

Reply:
In addition to what Matt wrote, the broadcasting should be tested (if we keep it). It's not clear to me line 574 will work if one of the objects only has one time.

Reply:
or just if flag anyway and size match do it

Reply:
Ack, I was in the middle of a review when the other comments appeared, so this is a little out of order.
But I absolutely agree with Matt that the default should be that data with a time (or frequency) label should only be applied to data with the same label. Keywords to override this is fine, but the general user would expect labels to match.

Reply:
Okay, so to compromise, what I'll do is just check the arrays match within the tolerance in the `UVFlag` in the case that `uvf.Nfreqs == uvd.Nfreqs` and raise an error if they don't, so that one can still flag in this scenario but they would need to manually raise `uvf._freq_array.tols` (so its kind of an advanced user kind of thing).

and the implicit broadcasting is covered: last test in the test function

Reply:
Okay that will work for me for now. :+1: